### PR TITLE
Add some more documentation to storage module

### PR DESCRIPTION
**Pull request by Hoverbear**
## What have you changed? (mandatory)

Adds some additional documentation to `storage`, particularly around the `Command` and `async_*` commands.

Future PRs will add more, but I wanted to keep the size small and reviewable.

## What are the type of the changes? (mandatory)

Just docs!

## How has this PR been tested? (mandatory)

I ran the CI!

## Does this PR affect documentation (docs) or release note? (mandatory)

Nope

## Does this PR affect tidb-ansible update? (mandatory)

Nope

## Refer to a related PR or issue link (optional)

Nope

## Benchmark result if necessary (optional)

Nope

## Add a few positive/negative examples (optional)

Nope

### CODE REVIEW

File: src/storage/mod.rs [hunk 1]
```diff
@@ -86,80 +98,165 @@ pub enum StorageCb {
...
+        /// will be rolled back.  `"k3"` will not be affected, because its lock_ts is not contained in
+        /// `txn_status`. `"k4"` will not be affected either, because it doesn't have a non-committed
+        /// version.
         txn_status: HashMap<u64, u64>,
         scan_key: Option<Key>,
         key_locks: Vec<(Key, Lock)>,
     },
+    /// Resolve locks on `resolve_keys` according to `start_ts` and `commit_ts`.
+    ///
+    /// During the GC operation, this should be called to clean up stale locks whose timestamp is
```
Comment (MyonKeminta) [inline src/storage/mod.rs]: @zhangjinpeng1987 If I didn't got into mistake, I think this is not used for GC? PTAL

File: src/storage/mvcc/mod.rs [hunk 2]
```diff
@@ -1,5 +1,7 @@
 // Copyright 2016 TiKV Project Authors. Licensed under Apache-2.0.
 
+//! Multi-view concurrency control functionality.
```
Comment (lonng) [inline src/storage/mvcc/mod.rs]: `Multi-Version concurrency control functionality`?

File: src/storage/mod.rs (hunk 1 above)
Comment (zhangjinpeng1987) [reply src/storage/mod.rs]: Yes, it is triggered by read request.

File: src/storage/mod.rs [hunk 3]
```diff
@@ -58,12 +65,17 @@ pub fn is_short_value(value: &[u8]) -> bool {
...
 #[derive(Debug, Clone)]
 pub enum Mutation {
+    /// Put `Value` into `Key`, overwriting any existing value.
     Put((Key, Value)),
+    /// Delete `Key`.
     Delete(Key),
+    /// Set a lock on `Key`.
     Lock(Key),
-    Insert((Key, Value)), // has a constraint that key should not exist.
+    /// Put `Value` into `Key` if `Key` does not yet exist.
```
Comment (zhangjinpeng1987) [inline src/storage/mod.rs]: Return "Should Not Exist" error when the key exist.

File: src/storage/mod.rs [hunk 4]
```diff
@@ -86,80 +98,165 @@ pub enum StorageCb {
...
         keys: Vec<(Key, bool)>,
+        /// The primary lock. Secondary locks (from `keys`) will refer to the primary lock.
         primary: Vec<u8>,
+        /// The transaction timestamp.
         start_ts: u64,
         options: Options,
     },
+    /// Commit the transaction that started at `lock_ts`.
+    ///
+    /// This should be following a [`Prewrite`](Command::Prewrite) or [`AcquirePessimisticLock`](Command::AcquirePessimisticLock).
```
Comment (zhangjinpeng1987) [inline src/storage/mod.rs]: AcquirePessimisticLock's next step is prewrite not commit.

File: src/storage/mod.rs (hunk 1 above)
Comment (Hoverbear) [reply src/storage/mod.rs]: Removed this comment.

File: src/storage/mvcc/mod.rs (hunk 2 above)
Comment (Hoverbear) [reply src/storage/mvcc/mod.rs]: How embarrassing, thanks!

File: src/storage/kv/rocksdb_engine.rs [hunk 5]
```diff
@@ -72,6 +72,7 @@ impl Drop for RocksEngineCore {
     }
 }
 
+/// **Test functionality:** The RocksEngine is based on `RocksDB`.
```
Comment (brson) [inline src/storage/kv/rocksdb_engine.rs]: It's not obvious to me what `**Test functionality:**` means, and how the text after the colon relates to it. I would guess that it means this type is only for testing, though the name of the type doesn't indicate so.

File: src/storage/mvcc/mod.rs (hunk 2 above)
Comment (brson) [inline src/storage/mvcc/mod.rs]:
```suggestion
//! Multi-version concurrency control.
```

File: src/storage/mvcc/reader/reader.rs [hunk 6]
```diff
@@ -321,7 +321,7 @@ impl<S: Snapshot> MvccReader<S> {
         IterOption::new(l_bound, u_bound, true)
     }
 
-    // Return the first committed key which start_ts equals to ts
+    /// Return the first committed key which start_ts equals to ts
```
Comment (brson) [inline src/storage/mvcc/reader/reader.rs]:
```suggestion
    /// Return the first committed key for which `start_ts` equals `ts`
```

File: src/storage/txn/mod.rs [hunk 7]
```diff
@@ -1,5 +1,7 @@
 // Copyright 2016 TiKV Project Authors. Licensed under Apache-2.0.
 
+//! Storage Transaction functionality.
```
Comment (brson) [inline src/storage/txn/mod.rs]:
```suggestion
//! Storage transactions
```

File: src/storage/txn/process.rs [hunk 8]
```diff
@@ -451,7 +451,8 @@ fn process_read_impl<E: Engine>(
                 Ok(ProcessResult::Res)
             } else {
                 let next_scan_key = if has_remain {
-                    // There might be more locks.
+                    // Locks is resolved via drop (`_lock`) here.
```
Comment (brson) [inline src/storage/txn/process.rs]:
```suggestion
                    // Locks are released via drop (`_lock`) here.
```

Comment (MyonKeminta) [inline src/storage/txn/process.rs]: Does this means the `ResolveLock` operations resolves locks that were scanned from the db when dropping `_lock`?
This seems to be not true. `ResolveLock` is processed cycle by cycle, and in each cycle it goes to `process_read` (here), scan a set of locks, and then goes to `process_write` to delete these scanned locks. `scan_key` marks where the next cycle should start from, and `_lock` is not useful here so it is dropped.

File: src/storage/kv/rocksdb_engine.rs (hunk 5 above)
Comment (Hoverbear) [reply src/storage/kv/rocksdb_engine.rs]: Clarified, thanks!

File: src/storage/txn/process.rs (hunk 8 above)
Comment (Hoverbear) [reply src/storage/txn/process.rs]: Oh, okay. Thank you for clarifying.

File: src/storage/mod.rs [hunk 9]
```diff
@@ -86,80 +100,159 @@ pub enum StorageCb {
...
+        /// The key to start from. (`None` means start from the very beginning.)
         start_key: Option<Key>,
+        /// The result limit.
         limit: usize,
     },
+    /// Resolve locks according to `txn_status`.
+    ///
+    /// During the GC operation, this should be called to clean up stale locks whose timestamp is
+    /// before safe point.
+    ///
```
Comment (MyonKeminta) [inline src/storage/mod.rs]: It seems that this empty line is not necessary?

**Comment by Hoverbear** [comment]
@MyonKeminta Oh no! Can you help me understand the confusion?

**Comment by dcalvin** [comment]
> I used to be told to use third person when a function's comment starts with a verb, but now I have no idea about that...
> Apart from that, this PR LGTM.

Yes. In[ Code Comment Style ](https://github.com/tikv/tikv/blob/master/CODE_COMMENT_STYLE.md) we do have this suggestion to use a descriptive way to comment. When a sentence starts with a verb, we recommend using the 3rd person singular form of the verb to indicate that the command/line/function **does** something... But obviously, this is not something that we can enforce well (people have preferences in language).  I think a better way is to start a comment sentence with the **noun** to be clearer with no ambiguity @MyonKeminta @Hoverbear

**Comment by MyonKeminta** [comment]
@dcalvin Actually we can see that starting the comments with verbs' 3rd person form is accepted in Rust's official documents ([for example](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.with_capacity)). So personally I prefer following how they does.

**Comment by dcalvin** [comment]
> ow

w w w w w w w w w

**Comment by dcalvin** [comment]
> @dcalvin Actually we can see that starting the comments with verbs' 3rd person form is accepted in Rust's official documents ([for example](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.with_capacity)). So personally I prefer following how they does.
This makes sense to me. @nrc @brson @Hoverbear comments?

**Comment by Hoverbear** [comment]
@MyonKeminta Great. I reflected what I *think* you were referring to, if I was wrong could you point it out inline so I can know?

**Comment by MyonKeminta** [comment]
Thank you! What we are talking about is, for example,
```rust
    /// Get values of a set of keys in a batch from the snapshot.
    ///
    /// Only writes that are committed before `start_ts` are visible.
    pub fn async_batch_get(
```
The comment of the function starts with a verb "Get", whose subject is the function below. In this case the third person form is preferred, so the "Get" should be written as "Gets".

Maybe we can merge this PR first since it's already approved, and change these verbs later if it's necessary.

**Comment by MyonKeminta** [comment]
Hi! How's it going about this PR? May we merge it?

**Comment by brson** [comment]
+1. Get it merged.

**Pull request merged**